cmd/samConsensus/samConsensus.go

```diff
@@ -19,149 +18,99 @@ import (
+			if skipGappedConsensus { // ignore gapped position
+				continue
+			} else { // ignore gapped reads
+				p.CountF[dna.Gap] = 0
+				p.CountR[dna.Gap] = 0
+				consensusBase = maxBase(p)
 			}
-			ref[i].Seq[k] = current
+		}
+		if p.CountF[consensusBase] == 0 && p.CountR[consensusBase] == 0 { // skip if no data present
+			continue
+		}
+		refName = header.Chroms[p.RefIdx].Name
+		refBase = dna.ToUpper(refMap[refName][p.Pos-1])
+		refMap[header.Chroms[p.RefIdx].Name][p.Pos-1] = consensusBase
+		if refBase != consensusBase {
+			_, err = fmt.Fprintf(outVcfFile, "%s\t%d\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", refName, int64(p.Pos), ".", dna.BaseToString(refBase), dna.BaseToString(consensusBase), ".", ".", ".", ".")
+			exception.PanicOnErr(err)
 		}
 	}
+
 	err = outVcfFile.Close()
 	exception.PanicOnErr(err)
 
 	fasta.Write(outFile, ref)
 }
 
-func voter(v voteBase, maxList []dna.Base) dna.Base {
-	if v.A+v.G+v.T+v.C+v.Gap == 0 {
-		return dna.N
-	}
-
-	var max int32 = v.A
+func maxBase(p sam.Pile) dna.Base {
+	var max int = p.CountF[dna.A] + p.CountR[dna.A]
 	var outBase dna.Base = dna.A
 
-	if v.G > max {
-		max = v.G
+	if p.CountF[dna.G]+p.CountR[dna.G] > max {
+		max = p.CountF[dna.G] + p.CountR[dna.G]
 		outBase = dna.G
 	}
 
-	if v.C > max {
-		max = v.C
+	if p.CountF[dna.C]+p.CountR[dna.C] > max {
+		max = p.CountF[dna.C] + p.CountR[dna.C]
 		outBase = dna.C
 	}
 
-	if v.T > max {
-		max = v.T
+	if p.CountF[dna.T]+p.CountR[dna.T] > max {
+		max = p.CountF[dna.T] + p.CountR[dna.T]
 		outBase = dna.T
 	}
 
-	if v.Gap > max {
-		max = v.Gap
+	if p.CountF[dna.Gap]+p.CountR[dna.Gap] > max {
+		max = p.CountF[dna.Gap] + p.CountR[dna.Gap]
 		outBase = dna.Gap
 	}
 
 	//now we check for ties
-	var maxCount int32 = 0
-
-	if v.A == max {
-		maxCount++
-		maxList = append(maxList, dna.A)
+	candidateBases := make([]int, 0, 5)
+	if p.CountF[dna.A]+p.CountR[dna.A] == max {
+		candidateBases = append(candidateBases, int(dna.A))
 	}
-	if v.C == max {
-		maxCount++
-		maxList = append(maxList, dna.C)
+	if p.CountF[dna.C]+p.CountR[dna.C] == max {
+		candidateBases = append(candidateBases, int(dna.C))
 	}
-	if v.T == max {
-		maxCount++
-		maxList = append(maxList, dna.T)
+	if p.CountF[dna.T]+p.CountR[dna.T] == max {
+		candidateBases = append(candidateBases, int(dna.T))
 	}
-	if v.G == max {
-		maxCount++
-		maxList = append(maxList, dna.G)
+	if p.CountF[dna.G]+p.CountR[dna.G] == max {
+		candidateBases = append(candidateBases, int(dna.G))
 	}
-	if v.Gap == max {
-		maxCount++
-		maxList = append(maxList, dna.Gap)
+	if p.CountF[dna.Gap]+p.CountR[dna.Gap] == max {
+		candidateBases = append(candidateBases, int(dna.Gap))
 	}
 
-	if maxCount < 2 {
+	if len(candidateBases) == 1 {
 		return outBase
 	}
 
-	return maxList[numbers.RandIntInRange(0, len(maxList))]
-}
-
-type voteBase struct {
-	A   int32
-	C   int32
-	G   int32
-	T   int32
-	Gap int32
+	return dna.Base(candidateBases[numbers.RandIntInRange(0, len(candidateBases))])
```

Comment:
Why not have candidate bases as []dna.Base instead of int?  I see us turning the bases into ints and then turning them back into bases when they come out.

Reply:
True. fixed.